lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,299 @@
+            DatasetTerminalStateViolation: dataset is in terminal state and
+                can't be advanced
+            DatasetBadStateTransition: dataset cannot be advanced to the
+                specified state
+
+        Returns:
+            Dataset: a dataset object in the desired state
+        """
+        try:
+            dataset = (
+                Database.db_session.query(Dataset)
+                .filter_by(controller=controller, name=name)
+                .first()
+            )
+        except SQLAlchemyError as e:
+            raise DatasetSqlError(f"SQL error: {e}")
+        else:
+            if dataset is None:
+                raise DatasetNotFound(f"No dataset {controller}:{name}")
+        dataset.advance(state)
+        return dataset
+
+    @staticmethod
+    def attach_or_create(controller, name, state):
+        """
+        attach Attempt to fetch dataset state for the controller and dataset
+        name. If found, attempt to advance the state of the dataset to the
+        specified state value.
+
+        If the dataset does not currently exist in the state database, we'll
+        create a new one with the specified state value.
+
+        NOTE: In "steady state" we should always find a dataset tracker in
+        the database, and failure is a serious Pbench server error; TBD how
+        we want to handle that. Right now, the dataset tracker DB is empty
+        and since we only populate when taking in a new dataset through the
+        PUT API, this provides a mechanism to import new datasets into the
+        DB dynamically. Another option would be to let the audit report
+        generator do this.
+
+        Args:
+            controller (string): controller name
+            name (string): dataset name
+            state (States ENUM): desired new state for dataset
+
+        Raises:
+            DatasetSqlError: problem interacting with Database
+            DatasetTerminalStateViolation: dataset is in terminal state and
+                can't be advanced
+            DatasetBadStateTransition: dataset cannot be advanced to the
+                specified state
+
+        Returns:
+            Dataset: a dataset object in the desired state
+        """
+        try:
+            dataset = Dataset.attach(controller, name, state)
+        except DatasetNotFound:
+            # TODO: in the long term we will always take in datasets
+            # through authenticated PUT and there'll always be a
+            # tracker created then, which we'd expect to find. This
+            # is intended as a transition, so it's probably safe to
+            # assume any dataset we can't find is unauthenticated and
+            # unowned.
+            dataset = Dataset(controller=controller, name=name, state=state)
+            dataset.add()
+        else:
+            dataset.advance(state)
+        return dataset
+
+    @staticmethod
+    def get_name(filename):
```

Comment:
Why have this method at all?  Why not just translate the `filename` to a `name` automatically in all the interfaces instead of requiring the caller to do it?

Reply:
I didn't need it for my first caller (indexing). I did for the second, but wasn't sure I wanted to add the overhead unconditionally for everyone. I'm definitely going to do this somewhere, though; unpack is probably going to pass the entire path and my `pbench-state-manager.py` command will have to break it down into controller and name; I'm not going to add that in bash since it doesn't already do it.

I'm debating with myself (I don't agree yet) whether to do that at the top level or push it all the way down.

As I mentioned on our call this morning, one thought I had was to change these methods to all kwargs like the SQLAlchemy-based `__init__` and add a `path` parameter that will be used to generate the name if not already specified.